phenotrex/ml/shap_handler.py

```diff
@@ -0,0 +1,248 @@
+        """
+        try:
+            X_agg = self._used_features.astype(float)
+            shap_agg = self._used_shaps.astype(float)
+        except (ValueError, AttributeError) as e:
+            raise RuntimeError('No explanations saved.')
+        sample_names = self._sample_names
+        return X_agg, shap_agg, sample_names
+
+    def _get_sorted_by_shap_data(self, sort_by_idx=None) -> Tuple[
+        np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Sort features by absolute magnitude of shap values,
+        and return sorted features, shap values and feature names.
+
+        :param sort_by_idx: if an index into the sample names is passed,
+                            sorting will be based only on this sample's SHAP values.
+        :return: Used features, used shaps, and the feature names all sorted by absolute magnitude of shap value.
+        """
+        X_agg, shap_agg, _ = self._get_feature_data()
+
+        if shap_agg.ndim == 2:
+            if sort_by_idx is None:
+                feature_sort_inds = np.argsort(np.sum(np.abs(shap_agg), axis=0))[::-1]
+            else:
+                feature_sort_inds = np.argsort(np.abs(shap_agg[sort_by_idx, :]))[::-1]
+        else:
+            feature_axis = shap_agg.ndim - 1
+            nonfeature_axes = list(range(feature_axis))
+            absshap = np.apply_along_axis(np.abs, feature_axis, shap_agg)
+            if sort_by_idx is None:  # sort features by absolute change in shap over all classes and samples
+                sort_criterion = np.apply_over_axes(np.sum, absshap, nonfeature_axes)
+                feature_sort_inds = np.squeeze(np.argsort(sort_criterion))[::-1]
+            else:  # sort features by absolute change in shap over all classes for given sample idx
+                sort_criterion = np.sum(absshap[sort_by_idx, ...], axis=0)
+            feature_sort_inds = np.squeeze(np.argsort(sort_criterion))[::-1]
+        return X_agg[:, feature_sort_inds], shap_agg[..., feature_sort_inds], \
+               self._used_feature_names[feature_sort_inds]
+
+    def plot_shap_force(self, sample_name: str, **kwargs):
+        """
+        Create force plot of the sample associated with the given sample name.
+
+        :param sample_name:
+        :param out_file_path:
+        :param kwargs: additional keyword arguments passed on to `shap.force_plot()`
+        :return:
+        """
+        counts, shaps, sample_names = self._get_feature_data()
+        i = self._get_sample_index_with_name(sample_name)
+
+        sf, ss = counts[i, ...], shaps[i, ...]
+        if ss.ndim == 2:
+            ss = list(ss)
+        else:
+            ss = [ss, ]
+        class_names = self._class_names
+        for c, s in zip(class_names, ss):
+            shap.force_plot(base_value=self._shap_base_value, shap_values=s, features=sf,
+                            feature_names=self._used_feature_names, matplotlib=True, show=False,
+                            **kwargs)
```

Comment:
Here the short variable names should be changed. They are hard to read and error-prone: Is `c` even used inside the loop?

Reply:
very true, that's a remnant of where I used the class before where I had multiclass models. Fixed.